research/lamp-automated-model-parallelism/unet_pipe.py

```diff
@@ -0,0 +1,171 @@
+from monai.networks.blocks import Convolution, UpSample
+from monai.networks.layers.factories import Act, Conv, Norm
+from torch import nn
+from torchgpipe.skip import Namespace, pop, skippable, stash
+
+
+@skippable(stash=["skip"], pop=[])
+class Stash(nn.Module):
+    def forward(self, input: torch.Tensor):
+        yield stash("skip", input)
+        return input  # noqa  using yield together with return
+
+
+@skippable(stash=[], pop=["skip"])
+class PopCat(nn.Module):
+    def forward(self, input: torch.Tensor):
+        skip = yield pop("skip")
+        if skip is not None:
+            input = torch.cat([input, skip], dim=1)
+        return input
+
+
+def flatten_sequential(module: nn.Sequential):
+    """
+    Recursively make all the submodules sequential.
+
+    Args:
+        module: a torch sequential model.
+    """
+    if not isinstance(module, nn.Sequential):
+        raise TypeError("module must be a nn.Sequential instance.")
+
+    def _flatten(module):
+        for name, child in module.named_children():
+            if isinstance(child, nn.Sequential):
+                for sub_name, sub_child in _flatten(child):
+                    yield f"{name}_{sub_name}", sub_child
+            else:
+                yield name, child
+
+    return nn.Sequential(OrderedDict(_flatten(module)))
+
+
+class DoubleConv(nn.Module):
+    def __init__(
+        self,
+        spatial_dims,
+        in_channels,
+        out_channels,
+        stride=2,
+        act_1=Act.LEAKYRELU,
+        norm_1=Norm.BATCH,
+        act_2=Act.LEAKYRELU,
+        norm_2=Norm.BATCH,
+        conv_only=True,
+    ):
+        """
+        A sequence of Conv_1 + Norm_1 + Act_1 + Conv_2 (+ Norm_2 + Act_2).
+
+        `norm_2` and `act_2` are ignored when `conv_only` is True.
+        `stride` is for `Conv_1`, typically stride=2 for 2x spatial downsampling.
+
+        Args:
+            spatial_dims: number of the input spatial dimension.
+            in_channels: number of input channels.
+            out_channels: number of output channels.
+            stride: stride of the first conv., mainly used for 2x downsampling when stride=2.
+            act_1: activation type of the first convolution.
+            norm_1: normalization type of the first convolution.
+            act_2: activation type of the second convolution.
+            norm_2: normalization type of the second convolution.
+            conv_only: whether the second conv is convolution layer only. Default to True,
+                indicates that `act_2` and `norm_2` are not in use.
+        """
+        super(DoubleConv, self).__init__()
+        self.conv = nn.Sequential(
+            Convolution(spatial_dims, in_channels, out_channels, strides=stride, act=act_1, norm=norm_1, bias=False,),
+            Convolution(spatial_dims, out_channels, out_channels, act=act_2, norm=norm_2, conv_only=conv_only),
+        )
+
+    def forward(self, x):
+        return self.conv(x)
+
+
+class UNet(nn.Sequential):
```

Comment:
As this UNet implementation is confusing with our internal UNet, could you please help change the class name to some name specific to this research?